File: tests/test_indexing.cpp
#include <filesystem>

#include <catch2/catch_test_macros.hpp>
#include <highfive/H5File.hpp>
#include <mpi.h>

#include "index/index.h"

namespace fs = std::filesystem;

const size_t NNODES = 10;
const size_t SOURCE_OFFSET = 90;
const char * GROUP = "data";

class MPIFixture {
  public:
    MPIFixture() {
        int init;
        MPI_Initialized(&init);
        if (!init) {
            MPI_Init(nullptr, nullptr);
        }
        MPI_Comm_size(MPI_COMM_WORLD, &size);
        MPI_Comm_rank(MPI_COMM_WORLD, &rank);
    }

    ~MPIFixture() {
        int finalized;
        MPI_Finalized(&finalized);
        if (!finalized) {
            MPI_Finalize();
        }
    }

    int size;
    int rank;
};

#define CATCH_CONFIG_RUNNER

void generate_data(const fs::path& base) {
    std::vector<uint64_t> source_ids;
    std::vector<uint64_t> target_ids;
    source_ids.reserve(NNODES * NNODES);
    target_ids.reserve(NNODES * NNODES);

    for (size_t i = 0; i < NNODES; ++i) {
        for (size_t j = 0; j < NNODES; ++j) {
            source_ids.push_back(SOURCE_OFFSET + i);
            target_ids.push_back(j);
        }
    }

    HighFive::File file(base, HighFive::File::Overwrite);
    auto g = file.createGroup(GROUP);
    g.createDataSet("source_node_id", source_ids);
    g.createDataSet("target_node_id", target_ids);
    indexing::write(g, SOURCE_OFFSET + NNODES, NNODES);
}

int main(int argc, char* argv[]) {
    MPIFixture mpi;
    int result = Catch::Session().run(argc, argv);
    return result;
}

TEST_CASE("Indexing") {

    SECTION("Generate data") {
        generate_data("index_test.h5");
    }
    HighFive::File f("index_test.h5");
    auto g = f.getGroup(GROUP);

    std::vector<std::array<uint64_t, 2>> source_ranges;
    std::vector<std::array<uint64_t, 2>> source_edges;
    std::vector<std::array<uint64_t, 2>> target_ranges;
    std::vector<std::array<uint64_t, 2>> target_edges;

    auto gidx = g.getGroup("indices");
    auto sidx = gidx.getGroup("source_to_target");
    auto tidx = gidx.getGroup("target_to_source");
    REQUIRE_NOTHROW(sidx.getDataSet("node_id_to_ranges").read(source_ranges));
    REQUIRE_NOTHROW(sidx.getDataSet("range_to_edge_id").read(source_edges));
    REQUIRE_NOTHROW(tidx.getDataSet("node_id_to_ranges").read(target_ranges));
    REQUIRE_NOTHROW(tidx.getDataSet("range_to_edge_id").read(target_edges));

    SECTION("Read Index") {
        REQUIRE(source_ranges.size() == SOURCE_OFFSET + NNODES);
        REQUIRE(target_ranges.size() == NNODES);

        REQUIRE(source_edges.size() == NNODES);
        REQUIRE(target_edges.size() == NNODES * NNODES);
    }

    SECTION("Check source index") {
        for (size_t i = 0; i < SOURCE_OFFSET; ++i) {
            REQUIRE(source_ranges[i][0] == 0);
            REQUIRE(source_ranges[i][1] == 0);
        }

        for (size_t i = 0; i < NNODES; ++i) {
            REQUIRE(source_ranges[SOURCE_OFFSET + i][0] == i);
            REQUIRE(source_ranges[SOURCE_OFFSET + i][1] == i + 1);

            REQUIRE(source_edges[i][0] == NNODES * i);
            REQUIRE(source_edges[i][1] == NNODES * (i + 1));
        }
    }

    SECTION("Check target index") {
        for (size_t i = 0; i < NNODES; ++i) {
            REQUIRE(target_ranges[i][0] == NNODES * i);
            REQUIRE(target_ranges[i][1] == NNODES * (i + 1));

            // Target node i is connected to every jth source node
            for (size_t j = 0; j < NNODES; ++j) {
                REQUIRE(target_edges[NNODES * i + j][0] == NNODES * j + i);
                REQUIRE(target_edges[NNODES * i + j][1] == NNODES * j + i + 1);
            }
        }
    }
}

File: tests/test_indexing.py
import pytest
import h5py
import numpy as np
from mpi4py import MPI
import index_writer_py

NNODES = 10
SOURCE_OFFSET = 90
GROUP = "data"

@pytest.fixture(scope="module")
def mpi_comm():
    return MPI.COMM_WORLD

def generate_data(base, rank, comm):
    if rank == 0:
        source_ids = np.repeat(np.arange(SOURCE_OFFSET, SOURCE_OFFSET + NNODES), NNODES)
        target_ids = np.tile(np.arange(NNODES), NNODES)

        with h5py.File(base, 'w') as file:
            g = file.create_group(GROUP)
            g.create_dataset("source_node_id", data=source_ids)
            g.create_dataset("target_node_id", data=target_ids)

        # Verify the file exists and contains the expected datasets
        with h5py.File(base, 'r') as file:
            g = file[GROUP]
            assert "source_node_id" in g, "source_node_id dataset not found"
            assert "target_node_id" in g, "target_node_id dataset not found"
    
    comm.Barrier()

def test_indexing(mpi_comm):
    rank = mpi_comm.Get_rank()
    base = "index_test.h5"

    # Step 1: Create and write to file using all ranks
    generate_data(base, rank, mpi_comm)

    # Step 2: Call index writer from all nodes
    index_writer_py.write(base, GROUP, SOURCE_OFFSET + NNODES, NNODES)

    # Ensure all processes have completed writing before verification
    mpi_comm.Barrier()

    # Step 3: Verify from all ranks
    with h5py.File(base, 'r') as f:
        g = f[GROUP]
        gidx = g['indices']
        sidx = gidx['source_to_target']
        tidx = gidx['target_to_source']

        source_ranges = sidx['node_id_to_ranges'][:]
        source_edges = sidx['range_to_edge_id'][:]
        target_ranges = tidx['node_id_to_ranges'][:]
        target_edges = tidx['range_to_edge_id'][:]

        # Check index sizes
        assert len(source_ranges) == SOURCE_OFFSET + NNODES
        assert len(target_ranges) == NNODES
        assert len(source_edges) == NNODES
        assert len(target_edges) == NNODES * NNODES

        # Check source index
        for i in range(SOURCE_OFFSET):
            assert source_ranges[i][0] == 0
            assert source_ranges[i][1] == 0

        for i in range(NNODES):
            assert source_ranges[SOURCE_OFFSET + i][0] == i
            assert source_ranges[SOURCE_OFFSET + i][1] == i + 1
            assert source_edges[i][0] == NNODES * i
            assert source_edges[i][1] == NNODES * (i + 1)

        # Check target index
        for i in range(NNODES):
            assert target_ranges[i][0] == NNODES * i
            assert target_ranges[i][1] == NNODES * (i + 1)

            for j in range(NNODES):
                assert target_edges[NNODES * i + j][0] == NNODES * j + i
                assert target_edges[NNODES * i + j][1] == NNODES * j + i + 1

    # Ensure all processes have completed verification
    mpi_comm.Barrier()
